File: 2017_gridGame/main.cpp
#include <vector>
#include <climits>
class Solution {
public:
    long long gridGame(std::vector<std::vector<int>>& grid) {
        int N = grid[0].size();
        
        std::vector<std::vector<long long>> prefixSum(2, std::vector<long long>(N));
        prefixSum[0][0] = grid[0][0];
        prefixSum[1][0] = grid[1][0];
        
        for(int i = 1;i < N; ++i) {
            prefixSum[0][i] = prefixSum[0][i-1] + grid[0][i];
            prefixSum[1][i] = prefixSum[1][i-1] + grid[1][i];
        }
        
        long long res = std::numeric_limits<long long>::max();
        long long row1, row2;
        for(int i = 0; i < N; ++i) {
            row1 = prefixSum[0][N-1] - prefixSum[0][i];
            row2 = i > 0 ? prefixSum[1][i-1] : 0;
            res = std::min(res,std::max(row1, row2));
        }
        return res;
    }
};

File: 2017_gridGame/main.py
from typing import List

class Solution:
    def gridGame(self, grid: List[List[int]]) -> int:
        N = len(grid[0])
        prefixSum1 = [val for val in grid[0]]
        prefixSum2 = [val for val in grid[1]]

        for i in range(1, N):
            prefixSum1[i] += prefixSum1[i-1]
            prefixSum2[i] += prefixSum2[i-1]
        
        res = float("inf")
        for i in range(N):
            row1 = prefixSum1[-1] - prefixSum1[i]
            row2 = prefixSum2[i-1] if i > 0 else 0
            robot2PathSum = max(row1, row2)
            res = min(res, robot2PathSum)

        return res
             
if __name__ == "__main__":
    print(Solution().gridGame([[1,3,1,15],[1,3,3,1]]))

